File: src/main/java/cz/wa/secretagent/game/utils/ProjectileMover.java
package cz.wa.secretagent.game.utils;

import cz.wa.secretagent.game.simulator.entity.projectile.ProjectileHitTypes;
import cz.wa.secretagent.game.utils.collision.CollidedEntity;
import cz.wa.secretagent.game.utils.collision.CollidedTile;
import cz.wa.secretagent.game.utils.collision.CollisionDescriptor;
import cz.wa.secretagent.game.utils.collision.ProjectileCollider;
import cz.wa.secretagent.world.entity.Entity;
import cz.wa.secretagent.world.entity.EntityType;
import cz.wa.secretagent.world.entity.agent.AgentEntity;
import cz.wa.secretagent.world.entity.projectile.ProjectileEntity;
import org.apache.commons.math3.geometry.euclidean.twod.Vector2D;
import secretAgent.world.SamWorld;

/**
 * Util class to move and collide projectile. 
 * 
 * @author Ondrej Milenovsky
 */
public class ProjectileMover {
    private final SamWorld world;
    private final ProjectileEntity projectile;
    private final double timeS;

    public ProjectileMover(SamWorld world, ProjectileEntity projectile, double timeS) {
        super();
        this.world = world;
        this.projectile = projectile;
        this.timeS = timeS;
    }

    /**
     * Move the bullet in the world and do all collisions.
     * @return true if the bullet is still alive
     */
    public boolean move() {
        ProjectileCollider projectileCollider = new ProjectileCollider(world, projectile, timeS);
        CollisionDescriptor collision = projectileCollider.findNearestCollision(
                ProjectileHitTypes.TILE_TYPES, ProjectileHitTypes.ENTITY_TYPES, timeS);

        if (collision != null) {
            // projectile collided
            if (collision instanceof CollidedTile) {
                // collided with tile
                // TODO impact effect
            } else if (collision instanceof CollidedEntity) {
                // collided with entity
                Entity entity = ((CollidedEntity) collision).getEntity();
                if (entity.getType() == EntityType.AGENT) {
                    new AgentHurter(world, (AgentEntity) entity).hit(projectile);
                }
            }
            world.getEntityMap().removeEntity(projectile);
            return false;
        } else {
            // no collision, just move the projectile
            double remainingDist = projectile.getRemainingDist();
            if (remainingDist <= 0) {
                // bullet expired
                world.getEntityMap().removeEntity(projectile);
                return false;
            }
            // move the projectile
            Vector2D moveVector = projectileCollider.getMoveVector();
            projectile.setRemainingDist(projectile.getRemainingDist() - projectile.getSpeed().getNorm()
                    * timeS);
            projectile.setPos(projectile.getPos().add(moveVector));
            return true;
        }
    }
}

File: src/main/kotlin/secretAgent/game/utils/ProjectileMover.kt
package secretAgent.game.utils

import secretAgent.game.simulator.entity.ProjectileHitTypes
import secretAgent.world.SamWorld
import secretAgent.world.entity.EntityType
import secretAgent.world.entity.ProjectileEntity
import secretAgent.world.entity.agent.AgentEntity

/**
 * Util class to move and collide projectile.
 *
 * @author Ondrej Milenovsky
 */
class ProjectileMover(private val world: SamWorld,
                      private val projectile: ProjectileEntity,
                      private val timeS: Double) {

    /**
     * Move the bullet in the world and do all collisions.
     * @return true if the bullet is still alive
     */
    fun move(): Boolean {
        val projectileCollider = ProjectileCollider(world, projectile, timeS)
        val collision = projectileCollider.findNearestCollision(
                ProjectileHitTypes.TILE_TYPES, ProjectileHitTypes.ENTITY_TYPES, timeS)

        if (collision != null) {
            // projectile collided
            if (collision is CollidedTile) {
                // collided with tile
                // TODO impact effect
            } else if (collision is CollidedEntity) {
                // collided with entity
                val entity = collision.entity
                if (entity.type == EntityType.AGENT)
                    AgentHurter(world, entity as AgentEntity).hit(projectile)
            }
            world.entityMap.removeEntity(projectile)
            return false
        } else {
            // no collision, just move the projectile
            val remainingDist = projectile.remainingDist
            if (remainingDist <= 0) {
                // bullet expired
                world.entityMap.removeEntity(projectile)
                return false
            }
            // move the projectile
            val moveVector = projectileCollider.moveVector
            projectile.remainingDist = projectile.remainingDist - projectile.speed.norm * timeS
            projectile.pos = projectile.pos.add(moveVector)
            return true
        }
    }
}